### PR TITLE
[HINTS]Add implementation for hints & utils for function nondet_bigint3 (cairo_secp/bigint.cairo )

### PR DESCRIPTION
**Pull request by pefontana**
# Add implementation for hints & utils of function nondet_bigint3 (cairo_secp/bigint.cairo)

## Description
* Add Secp::SecpUtils::split function. [Python Code](https://github.com/starkware-libs/cairo-lang/blob/167b28bcd940fd25ea3816204fa882a0b0a49603/src/starkware/cairo/common/cairo_secp/secp_utils.py#L12)
* Add MemorySegmentManager methods `write_arg` and `gen_arg_vec_bigint`. [Python code](https://github.com/starkware-libs/cairo-lang/blob/13cef109cd811474de114925ee61fd5ac84a25eb/src/starkware/cairo/lang/vm/memory_segments.py#L181)
* Add [hint](https://github.com/starkware-libs/cairo-lang/blob/167b28bcd940fd25ea3816204fa882a0b0a49603/src/starkware/cairo/common/cairo_secp/bigint.cairo#L65) implementation:
```
%{
    from starkware.cairo.common.cairo_secp.secp_utils import split
    segments.write_arg(ids.res.address_, split(value))
%}
```
This hint has no integration test because it needs to be executed after a hint that initializes the `value` variable in the VM scope.

## Checklist
- [ ] Linked to Github Issue
- [x] Unit tests added
- [x] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: src/vm/hints/secp/secp_utils.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    for _n in 0..3 {
+        (num, remainder) = num.clone().div_rem(&base);
+        canonical_repr.push(remainder);
+    }
+    if num != bigint!(0) {
+        return Err(VirtualMachineError::SecpSplitutOfRange(integer.clone()));
+    }
+
+    Ok(canonical_repr)
+}
```
Comment (Oppen) [inline src/vm/hints/secp/secp_utils.rs:30]:
```suggestion
pub fn split(integer: &BigInt) -> Result<[BigInt; 3], VirtualMachineError> {
    if integer.is_negative() {
        return Err(VirtualMachineError::SecpSplitNegative(integer.clone()));
    }
    let base = bigint!(1) << 86;
    let base_max = base - bigint!(1);
    let mut num = integer.clone();
    let mut canonical_repr: [Bigint; 3];
    for i in 0..3 {
        canonical_repr[i] = (&num & &base_max).to_owned();
        num >>= 86;
    }
    if !num.is_zero() {
        return Err(VirtualMachineError::SecpSplitutOfRange(integer.clone()));
    }

    Ok(canonical_repr)
}
```

File: src/vm/vm_memory/memory_segments.rs [hunk 2]
```diff
@@ -67,6 +71,40 @@ impl MemorySegmentManager {
         relocation_table.pop();
         Ok(relocation_table)
     }
+
+    pub fn gen_arg_vec_bigint(
+        &self,
+        arg: &[BigInt],
+        apply_modulo_to_args: bool,
+        prime: &BigInt,
```
Comment (Oppen) [inline src/vm/vm_memory/memory_segments.rs:79]:
```suggestion
        prime: Option<&BigInt>,
```

File: src/vm/vm_memory/memory_segments.rs [hunk 3]
```diff
@@ -67,6 +71,40 @@ impl MemorySegmentManager {
...
+        }
+    }
+
+    pub fn write_arg(
+        &mut self,
+        memory: &mut Memory,
+        ptr: &MaybeRelocatable,
+        arg: &dyn Any,
+        apply_modulo_to_args: bool,
+        prime: &BigInt,
```
Comment (Oppen) [inline src/vm/vm_memory/memory_segments.rs:99]:
```suggestion
        prime: Option<&BigInt>,
```

File: src/vm/vm_memory/memory_segments.rs [hunk 4]
```diff
@@ -67,6 +71,40 @@ impl MemorySegmentManager {
...
+            arg.iter()
+                .map(|x| MaybeRelocatable::from(x.clone()))
+                .collect()
+        }
+    }
+
+    pub fn write_arg(
+        &mut self,
+        memory: &mut Memory,
+        ptr: &MaybeRelocatable,
```
Comment (Oppen) [inline src/vm/vm_memory/memory_segments.rs:96]:
```suggestion
        ptr: &Relocatable,
```

**Review by Oppen — CHANGES_REQUESTED** [review]

File: src/vm/hints/secp/secp_utils.rs (hunk 1 above)
Comment (pefontana) [reply src/vm/hints/secp/secp_utils.rs]: I ended up doing https://github.com/lambdaclass/cleopatra_cairo/pull/322/commits/a69e0e95a6476a310a9dcf567f3489f936c2138b

**Comment by pefontana** [comment]
Thanks for the review @Oppen 
I made the corrections [here](https://github.com/lambdaclass/cleopatra_cairo/pull/322/commits/e46d1b9869b44f12db1d5a136052cffec3b352eb)

I will move this PR to draft, until PR #327 is merged

File: src/vm/hints/secp/bigint_utils.rs [hunk 5]
```diff
@@ -37,10 +37,10 @@ pub fn nondet_bigint3(
         }
     };
 
-    let arg = split(value)?;
+    let arg: Vec<BigInt> = split(value)?.to_vec();
 
     vm.segments
-        .write_arg(&mut vm.memory, &res_address, &arg, true, &vm.prime)
+        .write_arg(&mut vm.memory, &res_reloc, &arg, true, Some(&vm.prime))
```
Comment (Oppen) [inline src/vm/hints/secp/bigint_utils.rs:43]: Using the `Option` is supposed to make the boolean redundant:
```suggestion
        .write_arg(&mut vm.memory, &res_reloc, &arg, Some(&vm.prime))
```

File: src/vm/hints/secp/secp_utils.rs [hunk 6]
```diff
@@ -1,34 +1,48 @@
...
+Note that the limbs do not have to be in the range [0, BASE).
+prime should be the Cairo field, and it is used to handle negative values of the limbs.
+*/
+pub fn pack(d0: &BigInt, d1: &BigInt, d2: &BigInt, prime: &BigInt) -> BigInt {
+    let unreduced_big_int_3 = vec![d0, d1, d2];
+
+    unreduced_big_int_3
+        .iter()
+        .enumerate()
+        .map(|(idx, value)| as_int(value, prime) * (bigint!(1) << (idx * 86)))
```
Comment (Oppen) [inline src/vm/hints/secp/secp_utils.rs:42]: I missed a spot, product by the power of two is equivalent to making the shift only:
```suggestion
        .map(|(idx, value)| as_int(value, prime) << (idx * 86))
```

File: src/vm/vm_memory/memory_segments.rs [hunk 7]
```diff
@@ -67,6 +71,46 @@ impl MemorySegmentManager {
...
+                .iter()
+                .map(|x| MaybeRelocatable::from(x.mod_floor(prime)))
+                .collect());
+        } else {
+            Ok(arg
+                .iter()
+                .map(|x| MaybeRelocatable::from(x.clone()))
+                .collect())
+        }
+    }
```
Comment (Oppen) [inline src/vm/vm_memory/memory_segments.rs:93]:
```suggestion
    pub fn gen_arg_vec_bigint(
        &self,
        arg: &[BigInt],
        prime: Option<&BigInt>,
    ) -> Result<Vec<MaybeRelocatable>, MemoryError> {
        if let Some(ref prime) = prime {
            return Ok(arg
                .iter()
                .map(|x| MaybeRelocatable::from(x.mod_floor(prime)))
                .collect());
        } else {
            Ok(arg
                .iter()
                .map(|x| MaybeRelocatable::from(x.clone()))
                .collect())
        }
    }
```

File: src/vm/vm_memory/memory_segments.rs [hunk 8]
```diff
@@ -67,6 +71,46 @@ impl MemorySegmentManager {
...
+            let data = self.gen_arg_vec_bigint(vector, apply_modulo_to_args, prime)?;
+            self.load_data(
+                memory,
+                &MaybeRelocatable::from((ptr.segment_index, ptr.offset)),
+                data,
+            )
+        } else {
+            Err(MemoryError::WriteArg)
+        }
+    }
```
Comment (Oppen) [inline src/vm/vm_memory/memory_segments.rs:113]:
```suggestion
    pub fn write_arg(
        &mut self,
        memory: &mut Memory,
        ptr: &Relocatable,
        arg: &dyn Any,
        prime: Option<&BigInt>,
    ) -> Result<MaybeRelocatable, MemoryError> {
        if let Some(vector) = arg.downcast_ref::<Vec<BigInt>>() {
            let data = self.gen_arg_vec_bigint(vector, prime)?;
            self.load_data(
                memory,
                &MaybeRelocatable::from((ptr.segment_index, ptr.offset)),
                data,
            )
        } else {
            Err(MemoryError::WriteArg)
        }
    }
```

**Review by Oppen — CHANGES_REQUESTED** [review]

File: src/vm/hints/secp/bigint_utils.rs (hunk 5 above)
Comment (pefontana) [reply src/vm/hints/secp/bigint_utils.rs]: Yes, you are right, didn't catch that
https://github.com/lambdaclass/cleopatra_cairo/pull/322/commits/0b26d6d80b4980192a4c67e88249e0e8c60a3594

**Review by Oppen — APPROVED** [review]

**Pull request merged**